File: acmicpc/A12015.py
from sys import stdin as s


def solution(seq: list[int], n: int) -> None:
    q = [seq[0]]
    ans = 1
    for num in seq[1:]:
        if num <= q[-1]:
            low, high = 0, len(q) - 1
            while low < high:
                mid = (low + high) // 2
                if q[mid] >= num:
                    high = mid
                else:
                    low = mid + 1
            q[high] = num
        else:
            q.append(num)
            ans += 1
    print(ans)


def reader():
    n = int(s.readline().strip())
    seq = map(int, s.readline().split())
    solution(list(seq), n)


if __name__ == '__main__':
    reader()

File: acmicpc/A12015.java
package acmicpc;

import java.io.*;
import java.util.*;

public class A12015 {
    public static int solution(int[] seq, int length) {
        ArrayList<Integer> q = new ArrayList<>();
        q.add(seq[0]);
        int ans = 1;
        for (int i = 1; i < seq.length; i++) {
            int num = seq[i];
            if (num > q.get(q.size() - 1)) {
                q.add(num);
                ans++;
                continue;
            }
            int low = 0, high = q.size() - 1;
            while (low <= high) {
                int mid = low + (high - low) / 2;
                if (q.get(mid) < num)
                    low = mid + 1;
                else
                    high = mid - 1;
            }
            q.set(low, num);
        }
        return ans;
    }

    public void reader() throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int seq[] = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++)
            seq[i] = Integer.parseInt(st.nextToken());
        System.out.println(solution(seq, n));
    }

    public static void main(String[] args) throws NumberFormatException, IOException {
        A12015 test = new A12015();
        test.reader();
    }
}
